### DISCUSSION TITLE
Is it anti pattern to use Store inside a Context Provider?

### DISCUSSION THREAD

**Discussion by KrisLau** [pmndrs/zustand#1610] [General] [answered]
I need the reactive states from another store in order to show an accurate FlatList in my overlay so I created a context provider and consumed the baseball store there and then created another pitch actions store in order to display the `subItem`s. It works but is this anti-pattern and if so, how should I go about refactoring this?

Shortened version of my code:
```js
function PitchActionsProvider({children}) {
  const strikeCount = useBaseballStore(state => state.strikeCount);

  //region Pitch Actions
  const calledStrike = {
    icon: pitchIcons.calledStrike,
    name: 'Called Strike',
    isTriggerAnimation: true,
    animationTitle: 'Strike',
    onPress: () => incrementStrikeCount(),
  };
  const foulTipOut = {
    icon: pitchIcons.foulTipOut,
    name: 'Foul Tip Out',
    isShown: strikeCount === 2,
    onPress: () => incrementStrikeCount(),
  };
  const ballInPlay = {
    icon: pitchIcons.ballInPlay,
    name: 'Ball in Play',
    subItems: [groundBall, hardGroundBall, flyBall, lineDrive, bunt, popFly],
    style: {color: Colors.BLUE},
  };

  const allActions = [
    [ball, calledStrike, swingMiss, droppedThirdStrike],
    [foulBall, ballInPlay, foulTipOut],
    [
      hitByPitch,
      intentionalBall,
      intentionalWalk,
      catcherInterference,
      balk,
      batterOutOther,
      putOnBase,
    ],
  ];

  const usePitchActionsStore = create(set => ({
    pitchActions: allActions,
    actions: allActions,
    setActions: actions => set(() => ({actions: actions})),
    selections: [],
    setSelections: selections => set(() => ({selections: selections})),
    resetCurrentActions: () => set(state => ({actions: state.pitchActions})),
  }));

  function loadedBasesCount() {
    return [firstBaseRunner != null, secondBaseRunner != null, thirdBaseRunner != null].filter(
      Boolean,
    ).length;
  }

  const value = {
    usePitchActionsStore,
  };
  return <PitchActionsContext.Provider value={value}>{children}</PitchActionsContext.Provider>;
}
```

**Comment by dai-shi** [comment]
Creating a store in render is a bad idea. It will be re-created on the next render, and could lead to infinite loop.

(You could use `useRef` or something to make the store stable, in that case, `createStore` should be preferred instead of `create`.)

**Reply by KrisLau** [reply]
I tried wrapping it in a useRef but for some reason now the store doesn't seem to be reactive. It doesn't change the values. When I use `createStore` I get an error

**Reply by dai-shi** [reply]
Can you create a very very small example with https://codesandbox.io/s/react-new ?

**Reply by KrisLau** [reply]
I managed to get it working right now by moving the store to a separate file and calling an initialize method. Is this ok? If not, I'll get started on that repro!

Context Provider:
```js
function PitchActionsProvider({children}) {
  const initialize = usePitchActionsStore(state => state.initialize);
  const strikeCount = useBaseballStore(state => state.strikeCount);

  ...

  const allActions = [
    [ball, calledStrike, swingMiss, droppedThirdStrike],
    [foulBall, ballInPlay, foulTipOut],
    [
      hitByPitch,
      intentionalBall,
      intentionalWalk,
      catcherInterference,
      balk,
      batterOutOther,
      putOnBase,
    ],
  ];
  initialize(allActions);

  const value = {};
  return <PitchActionsContext.Provider value={value}>{children}</PitchActionsContext.Provider>;
}
```

----
`pitch-actions-store.js`:
```js
import {create} from 'zustand';

const usePitchActionsStore = create(set => ({
  pitchActions: [],
  actions: [],
  setActions: actions => set(() => ({actions: actions})),
  selections: [],
  setSelections: selections => set(() => ({selections: selections})),
  initialize: pitchActions => set(() => ({pitchActions: pitchActions, actions: pitchActions})),
  resetCurrentActions: () => set(state => ({actions: state.pitchActions})),
}));

export {usePitchActionsStore};
```

**Reply by dai-shi** [reply]
In this case, I don't think you need to use context in the first place.

**Reply by KrisLau** [reply]
Sorry I'm not quite sure how else to get the share the values of other states from the store with other states? cause many of the actions have overlapping `subItems` like 
```js
const ballInPlay = {
    icon: pitchIcons.ballInPlay,
    name: 'Ball in Play',
    subItems: [groundBall, hardGroundBall, flyBall, lineDrive, bunt, popFly], //this
    style: {color: Colors.BLUE},
  }; 
```
If you have any alternatives please let me know!

**Reply by dai-shi** [reply]
I may have misunderstood your use case. Can you provide much smaller example?

Back to the original question,

> Is it anti pattern to use Store inside a Context Provider?

No, it's a good pattern, but my recommendation is to pass a vanilla store via Provider, and use `useStore` hook in consumer components. Passing a hook through context or props can lead to violating react hooks rules.

**Reply by KrisLau** [reply]
Is this ok: https://snack.expo.dev/@kris_lau/mini-example?platform=ios? I tried to minimize the code as much as possible. Not sure why the `useEffect` doesnt work in the snack but it works in my project

**Reply by dai-shi** [reply]
I don't see any issues with zustand usage. It's still too big to understand a specific problem.
`useEffect` deps is wrong because it can cause infinite loop.
What would be the `value` you want to pass to the provider? Or, isn't it very important?

**Reply by KrisLau** [reply]
There's really no way to make it any smaller haha. Since you said it's ok the way it is, i guess it's there's no problem or do you think there's a better way to go about it?

**Reply by dai-shi** [reply]
Do you need all actions to explain your problem? Do you need style? Do you need multiple file? Just to try reducing lines.

If you think it's hard to make it smaller, do you want to draw a diagram of component tree and stores? https://excalidraw.com

**Reply by KrisLau** [reply]
@dai-shi Reduced it a little. 

Basically the only reason i'm using the provider is because if you look at `groundBall` and `single`, there are `subItems` which are references to other variables:
```js
  const single = {
    name: 'Single',
    onPress: () => {},
  };
  const groundBall = {
    name: 'Ground Ball',
    subItems: [single],  //<--- this
  }
```

and some which reference the baseball store:
```js
  const incrementBallCount = useBaseballStore(state => state.incrementBallCount);

  const ball = {
    name: 'Ball',
    onPress: () => incrementBallCount(), //<--- this
  };
```
---
Is there a way to have them initialized the same way as my current implementation with just the pitch actions store:
```js
// not valid code just trying to illustrate what I am trying to achieve
const usePitchActionsStore = create(set => ({
  single: {
    name: 'Single',
    onPress: () => {},
  };
  groundBall: {
    name: 'Ground Ball',
    subItems: [single],
  };
  ball: {
    name: 'Ball',
    onPress: () => baseballStore.incrementBallCount(),
  };
})
```
Or is my current implementation using the provider the best way to go around it? If not what do you suggest?

**Reply by dai-shi** [reply]
If you need to call an action of another store you can do it like this:
```js
const usePitchActionsStore = create(set => ({
  single: {
    name: 'Single',
    onPress: () => {},
  };
  groundBall: {
    name: 'Ground Ball',
    subItems: [single],
  };
  ball: {
    name: 'Ball',
    onPress: () => useBaseballStore.getState().incrementBallCount(),
  };
})
```

**Reply by KrisLau** [reply]
@dai-shi right but what if I need it to be reactive? not just the static state? As the ball, strike and out count increases, it changes what items are shown on the menu (I removed the property that references those to shorten the example but you can see it in my original post). Also what about state from the same store like the `subItems` in `groundBall` (there are dozens of references to same store state all over)?

**Reply by dai-shi** [reply]
It sounds like you want to put everything in a single store, which may be easier?

**Reply by KrisLau** [reply]
@dai-shi I have considered that but I still wouldn't know how to have the state within the store depend on another state in the same store so it would still result in a context unless there is a way to do so?

**Reply by dai-shi** [reply]
> to have the state within the store depend on another state in the same store

Oh, do you mean a kind of computed value?
```js
const useFooStore = create((set) => ({
  count: 0,
  inc: () => set((state) => ({ count: state.count + 1 })),
  doubledCount: /* something like this ? */ myself.count * 2,
}))
```

Otherwise, can you clarify what "depend" means?

**Reply by KrisLau** [reply]
@dai-shi like in my example: 
```js
const usePitchActionsStore = create(set => ({
  single: {
    name: 'Single',
    onPress: () => {},
    isShown: ballCount > 2, //<--- references "ballCount" from the baseballStore & I need it to be reactive
  };
  groundBall: {
    name: 'Ground Ball',
    subItems: [single], //<---this references the "single" state above (from the same store) & I need it to be reactive
    isShown: outCount === 1, 
  };
  ...
}))
```

**Reply by dai-shi** [reply]
1. If you define two stores, they are just isolated, and you can't connect each other. (only hooks can)
2. deriving state requires third-party libraries.

That said, I think you might want to try changing the mental model and don't include derived state in stores. If that doesn't work in your mental model, reach out some of third party libs.

**Comment by KrisLau** [comment, marked as answer]
https://github.com/pmndrs/zustand/discussions/1610#discussioncomment-5070953
